File: graph/topological-sort/[백준]2252. 줄 세우기/solution.cpp
#include <bits/stdc++.h>

using namespace std;

int n, m, a, b;
vector<int> adj[32001];
int degree[32001];
queue<int> q;

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    cin >> n >> m;
    while (m--) {
        cin >> a >> b;
        adj[a].push_back(b);
        degree[b]++;
    }

    for (int i = 1; i <= n; i++)
        if (!degree[i])
            q.push(i);

    while (!q.empty()) {
        int now = q.front(); q.pop();
        cout << now << ' ';
        for (int i = 0; i < adj[now].size(); i++) {
            int next = adj[now][i];
            if (--degree[next] == 0)
                q.push(next);
        }
    }
}

File: graph/topological-sort/[백준]2252. 줄 세우기/solution.py
import sys
from collections import deque
input = sys.stdin.readline

adj = [[] for _ in range(32001)]
degree = [0 for _ in range(32001)]
n, m = map(int, input().split())
for i in range(m):
    a, b = map(int, input().split())
    adj[a].append(b)
    degree[b] += 1

d = deque()
for i in range(1, n + 1):
    if degree[i] == 0:
        d.append(i)

while d:
    now = d.popleft()
    print(now, end=' ')
    for next in adj[now]:
        degree[next] -= 1
        if degree[next] == 0:
            d.append(next)